### ISSUE TITLE
Basic Questions

### ISSUE THREAD

**Issue by ennerf** [GSI-CS-CO/chart-fx#26]
## Background
First of all, thanks for writing a JavaFX charting library. I've been using `JChart2D` for what seems like forever, and I can't wait to finally get rid of the last remaining Swing parts. 

I've been working on an application that visualizes streaming sensor/measurement data in real-time. One of the features is a way for non-programmers to create custom charts via XML files where they can specify equations as well as things like default ranges, colors, styles, and so on. For example

```xml
<chart title="GYROSCOPE" min_range="-3" max_range="+3" range_policy="expanding">
    <trace label="gyroX" units="rad/s" color="red" style="solid" value="fbk.gyroX" />
    <trace label="gyroY" units="rad/s" color="green" style="solid" value="fbk.gyroY" />
    <trace label="gyroZ" units="rad/s" color="blue" style="solid" value="fbk.gyroZ" />
</chart>
```
![image](https://user-images.githubusercontent.com/5491587/64635485-15a90e00-d400-11e9-852c-d6956997f5fb.png)

and

```xml
<chart title="LATENCY" min_range="0" max_range="0.5" range_policy="expanding">
    <trace label="Round Trip Time (RTT)" units="ms" color="black" style="points" value="(fbk.pcRxTime - fbk.pcTxTime)*1E3" />
    <trace label="Hardware Response Time" units="ms" color="blue" style="points" value="(fbk.hwTxTime - fbk.hwRxTime)*1E3" />
    <trace label="Transmit Time dt" units="ms" color="green" style="points" value="isNaN(prevFbk.pcTxTime) ? 0 : (fbk.pcTxTime - prevFbk.pcTxTime)*1E3" />
</chart>
```

![image](https://user-images.githubusercontent.com/5491587/64635589-5012ab00-d400-11e9-832b-d3071903c96b.png)

Unfortunately, I keep running into various basic issues when trying to port the charts over to chart-fx.

## Questions / Problems

* We typically want to show the last N measurements, or the last (up to N) measurements within some time period. It'd be nice to have a `LimitedCircularDoubleDataSet`. It sounds very much like #8, so I'd second that as a feature request. Once I'm more familiar with this library I may be able to contribute something.

* How can I dynamically set colors for individual datasets? The only reference I've found was in [RollingBufferLegacyExample](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-samples/src/main/java/de/gsi/chart/samples/legacy/RollingBufferLegacySample.java#L117) via setting the chart style for `.default-color{index}.chart-series-line{...}`. Unfortunately, doing that doesn't seem to have any effect.

* Similarly, how can I selectively change the style of individual datasets? e.g. render one dataset with only markers (e.g. MATLAB's `.`, `o`, `x`), and another dataset as a solid/dashed/dotted line without markers (e.g. MATLAB's `-`, `--`, `:`). `DataSet::getDataStyleMap::put` seems to apply only to individual points

* How can I render a `dashed` line? I found some usages of `DashPatternStyle` in the `GridRenderer`, but found nothing about using it for individual datasets.

* What JDK/JFX version were you using for the performance comparisons in your paper? In particular, I'd be interested to know whether the numbers are before or after the [Marlin renderer](https://static.rainfocus.com/oracle/oow17/sess/1493054732447001bY8w/PF/javaone-marlin-talk_1507132140815001B9Fs.pdf) changes.

Thanks!

Edit: cleaned up the questions a bit

**Comment by ennerf** [comment]
Also, below is a list of some (probable) bugs I've encountered so far with the `XYChart`. They are minor, so I don't think it's worth creating new issues. Overall, the library seems quite nice 👍 

* Zooming in too far (e.g. >10x using the mouse area) crashes the entire runtime in a way that can't be recovered from

* Enabling/disabling grid lines does not trigger a redraw. Changes become visible after e.g. zooming in and out

* When zooming in, the Y-axis label eventually gets pushed outside of bounds. It may be worth hiding the label if it doesn't fit anymore. This is really nitpicky though and probably not worth the effort :)

normal ![image](https://user-images.githubusercontent.com/5491587/64653606-202ace00-d427-11e9-8c5f-325462529eca.png) zoomed-in ![image](https://user-images.githubusercontent.com/5491587/64653633-35076180-d427-11e9-86a9-3e565745b5eb.png)

**Comment by wirew0rm** [comment]
Hello Florian,

first of all, thanks for your interest in our charting framework and also for providing your user story, it's allways nice to see what other people are building.

> ```
> <chart title="GYROSCOPE" min_range="-3" max_range="+3" range_policy="expanding">
>     <trace label="gyroX" units="rad/s" color="red" style="solid" value="fbk.gyroX" />
>     <trace label="gyroY" units="rad/s" color="green" style="solid" value="fbk.gyroY" />
>     <trace label="gyroZ" units="rad/s" color="blue" style="solid" value="fbk.gyroZ" />
> </chart>
> ```

This actually looks very close to the fxml code for applications using chartfx. Currently most of our code manually instanciates the Chart components, so there are some attributes that cannot be set in a nice way from fxml, but we are open to improving that. e.g.
```
<XYChart fx:id="waveletChart">
  <plugins>
    <Zoomer />
    <Panner />
  </plugins>
  <renderers>
    <ContourDataSetRenderer />
  </renderers>
  <datasets>
    ....
  </datasets>
</XYChart>
```
Would give you a chart with zoom and panning functionality enabled and a renderer to render 3D contour data instead of the default xy data. Your traces would be DataSet implementations which expose all the properties you need to the fxml loader. Of course if you already have your xml format and parsing routines, this might not be a good approach, it just looked a lot like your code and I think we do not have any public fxml examples yet.

> ## Questions / Problems
> 
>     * We typically want to show the last N measurements, or the last (up to N) measurements within some time period. It'd be nice to have a `LimitedCircularDoubleDataSet`. It sounds very much like #8, so I'd second that as a feature request. Once I'm more familiar with this library I may be able to contribute something.

Did you check the `LimitedIndexedTreeDataSet` it should provide the functionality you described and for our use cases the performance was sufficient. I would suggest trying that and if it turns out to be a performance bottleneck, you can still switch to a custom DataSet implementation. Alternatively you could use the CircularDoubleDataSet and have some user code or a Plugin adjust your AxisRange to limit the data which is shown.
 
>     * How can I dynamically set colors for individual datasets? The only reference I've found was in [RollingBufferLegacyExample](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-samples/src/main/java/de/gsi/chart/samples/legacy/RollingBufferLegacySample.java#L117) via setting the chart style for `.default-color{index}.chart-series-line{...}`. Unfortunately, doing that doesn't seem to have any effect.
> 
>     * Similarly, how can I selectively change the style of individual datasets? e.g. render one dataset with only markers (e.g. MATLAB's `.`, `o`, `x`), and another dataset as a solid/dashed/dotted line without markers (e.g. MATLAB's `-`, `--`, `:`). `DataSet::getDataStyleMap::put` seems to apply only to individual points

CSS in JavaFX is complicated implementation wise, but in general most functionalities should work.
Note that markers are currently allways painted in strokeColor
```
dataSet.setStyle("strokeColor=pink; fillColor=green");
```
See [de/gsi/chart/XYChartCss.java](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-chart/src/main/java/de/gsi/chart/XYChartCss.java) for all supported properties.
The dataSetStyle map, as you correctly observed, allows to highlight single points in your data set by assigning a different style to them.

>     * How can I render a `dashed` line? I found some usages of `DashPatternStyle` in the `GridRenderer`, but found nothing about using it for individual datasets.

Dashed Lines are currently not supported, because the underlying JavaFX implementation has performance issues (even if you render on a canvas like we do it still allocates a Line Object for each dash). That's why we have our own dashing implementation for the grid renderer which uses solid lines with a dashed fill pattern. But this workaround only works correctly for horizontal and vertical lines.

>     * What JDK/JFX version were you using for the performance comparisons in your paper? In particular, I'd be interested to know whether the numbers are before or after the [Marlin renderer](https://static.rainfocus.com/oracle/oow17/sess/1493054732447001bY8w/PF/javaone-marlin-talk_1507132140815001B9Fs.pdf) changes.

The plot in the Paper was produced using JDK8, but main development nowadays happens on JDK11 and openJFX 12 and the relative Proportions have not changed significantly and speedup due to marlin was also moderate. I can rerun the performance evaluation overnight and post the results tomorrow.

**Comment by wirew0rm** [comment]
I also fixed all the issues from your second comment in our dev branch, except for the second one, which I put into the backlog.
The dev branch will become 11.1.0 soon.

Again, thanks for the detailed bug reports.

**Comment by ennerf** [comment]
Thank you for your quick feedback.

**FXML** 

That's good to know. Overall, I really like FXML and CSS. I usually work together with a UI designer, and the split makes it really easy to work with. The more that can be exposed, the better.

**LimitedIndexedTreeDataSet**

We often measure latency data that comes in at >=1KHz and is usually <=1 ms. The network latency is subject to GC, so I try really hard to reduce allocations to avoid even minor GCs (>=3ms) to not interfere with the results too much. The `LimitedIndexedTreeDataSet` adds a lot of extra objects, and the sorting is unnecessary given that the data is already sorted by time.

Writing a plugin is an interesting idea. I'll give that a try to start off with.

**Individual Styling**

Thanks. I was able to create individual styling using the CSS defines and a separate renderer for each dataset, as shown below. It don't know whether this is the intended usage, but it does work.

```Java
// Simplified dummy code
chart.getRenderers().clear();
for(Trace trace : traces) {
    // ...

    DoubleDataSet dataSet = new DoubleDataSet(name, x, y, x.length, false);
    
    // Define color/width via CSS styling of the data set
    String color = toCssColor(traceConfig.getColor());
    String style = ""
            + XYChartCss.MARKER_COLOR + "=" + color + ";"
            + XYChartCss.STROKE_COLOR + "=" + color + ";"
            + XYChartCss.FILL_COLOR + "=" + color + ";"
            + XYChartCss.STROKE_WIDTH + "=1;";
    if (traceConfig.getStyle() == TraceType.DASHED_LINE) {
        style += XYChartCss.STROKE_DASH_PATTERN + "=" + "5";
    }
    dataSet.setStyle(style);

    // Define line style via renderer
    ErrorDataSetRenderer traceRenderer = new ErrorDataSetRenderer();
    renderer.setDrawBubbles(false);
    renderer.setDrawBars(false);
    renderer.setDrawMarker(false);
    renderer.setPolyLineStyle(LineStyle.NONE);
    renderer.setErrorType(ErrorStyle.NONE);
    switch (type) {
            case SOLID_LINE:
            case DASHED_LINE:
                renderer.setPolyLineStyle(LineStyle.NORMAL);
                break;
            case INDIVIDUAL_POINTS:
                renderer.setDrawMarker(true);
                renderer.setMarker(DefaultMarker.RECTANGLE);
                renderer.setMarkerSize(1);
                break;
        }
   
    // Add to chart
    traceRenderer.getDatasets().add(dataSet);
    chart.getRenderers().add(traceRenderer);

}
```

**Comment by ennerf** [comment]
Btw. after having already ported parts of the application, I have to say that the zooming/panning features are really well done. That'll make a lot of users happy.

Feel free to close this issue unless you want to keep it open for reference.

**Comment by RalphSteinhagen** [comment]
@ennerf thanks for your interest and co-using our lib. 
>  It don't know whether this is the intended usage, but it does work.

Yes, this is one of the possible/intended usages. Our developers are quite diverse and favour a wide range from FXML/CSS-only up to fully programatic-style approaches. While the latter is (for us) easier to implement, we try to expand the FXML/CSS functionality where possible/reasonable/compatible with our other commitments (N.B. we actually build high-energy particle accelerators, having a charting lib is just one out of many tools for us).

**Comment by RalphSteinhagen** [comment]
> We often measure latency data that comes in at >=1KHz and is usually <=1 ms. The network latency is subject to GC, so I try really hard to reduce allocations and the number of minor GCs (>=3ms) whenever possible to not interfere with the results too much. The LimitedIndexedTreeDataSet adds a lot of extra objects, and the sorting is unnecessary given that the data is already sorted by time.

A bit of rationale behind [LimitedIndexedTreeDataSet](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/LimitedIndexedTreeDataSet.java): we specifically wrote this for a (perhaps) similar use-case where the sorting/tree was necessary due to:
* burst-conditions where the data arrives out-of-order (N.B. worker-thread-pool handling net-io)
* non-equidistant sampling (either DAQ errors or DAQ operating in triggered burst-mode).

Performance comparisons between a 'linear-list-based DataSets with post-process-sorting and dynamic resizing of the circular buffer queue' and the eventual `LimitedIndexedTreeDataSet` was similar/more favourable performance/maintainability for the latter under real-world conditions.

If you can assure equidistant sampling and sorted data, then [CircularDoubleErrorDataSet](https://github.com/GSI-CS-CO/chart-fx/blob/master/chartfx-dataset/src/main/java/de/gsi/dataset/spi/CircularDoubleErrorDataSet.java) might be the better choice for you.

_Out of engineering curiosity: if latencies <=1 ms and GCs are of concern, why do you use Java/JVM+GC in the first place. Many favour (ie. our real-time use case) either HW-based solutions and/or C++/Linux (+real-time extensions). Do you have documented public examples for your use-case/applications?_

_N.B. We are working on a twin-project to chart-fx in parallel (working title 'chart-qt' -- very imaginative, isn't it) that aims at a similar API/MVC-paradigm (notably DataSet structure) but implemented for our C++11/Qt environment._

**Comment by RalphSteinhagen** [comment]
I love that video... is this real-time ... if yes, we need this charting library.

regarding the issue: your observation is a valid point. :thumbsup: 
For the sake of keeping this issue short (and eventually close) could you repost this as a new issue? 

Maybe you could also add a video with the data point reduction disabled. This would help us to identify if it's the data reduction or line/marker rendering part (I presume the plot uses only markers, DataSet or DataSetError derived?). 

Thanks in advance.

**Comment by ennerf** [comment]
The thin clients make an interesting use case. I was already wondering what the use case for `BinarySerialiser` was. I deal quite a bit with serialization, and I wrote [MFL](https://github.com/HebiRobotics/MFL/) for creating `.mat` files, as well as a zero-allocation Java library for [Protobuf](https://developers.google.com/protocol-buffers/) (not open sourced yet, but I'm considering it). Protobuf in particular may be a better option than maintaining a custom binary format with versioning. These things go out of hand pretty quickly.

**Comment by ennerf** [comment]
I agree that this issue has already gotten way out of hand :)

Could you maybe add [Gitter](https://gitter.im/) to this repo? It's more like a chat and would make it a lot easier to handle small discussions. I have seen this work quite well for some repos, e.g., [gitter-channel](https://gitter.im/HdrHistogram/HdrHistogram?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) for [HdrHistogram](https://github.com/HdrHistogram/HdrHistogram)

**Comment by ennerf** [comment]
I believe all of the actual issues in this novel have been fixed. Closing.

**Comment by RalphSteinhagen** [comment]
:smile: @ebold-cscotos  maybe you could post another video of the final result. We'd love to have some external examples/references (N.B. especially since we are (re-)discussing the topic of UDP transmission internally). Thanks in any case.
